dask/array/core.py

```diff
@@ -2622,14 +2622,26 @@ def elemwise(op, *args, **kwargs):
         raise TypeError(msg % (op.__name__, str(sorted(set(kwargs) - set(['name', 'dtype'])))))
 
     args = [np.asarray(a) if isinstance(a, (list, tuple)) else a for a in args]
-
-    shapes = [getattr(arg, 'shape', ()) for arg in args]
-    shapes = [s if isinstance(s, Iterable) else () for s in shapes]
-    out_ndim = len(broadcast_shapes(*shapes))   # Raises ValueError if dimensions mismatch
-    expr_inds = tuple(range(out_ndim))[::-1]
-
     arrays = [a for a in args if not is_scalar_for_elemwise(a)]
     other = [(i, a) for i, a in enumerate(args) if is_scalar_for_elemwise(a)]
+    atop_args = concat((a, tuple(range(a.ndim)[::-1])) for a in arrays)
+
+    if (op == getitem and
+            isinstance(args[1], Array) and
+            _is_boolean_mask(args[0], args[1])):
+        # getitem has special rules for boolean masks utput shape is the same as the input
+        # on all dimensions except the one being indexed, where it's the number of Trues
+        # We only handle x[mask] here, so that's `args[0]`
+        # out_ndim = args[0].ndim - args[1].ndim + 1
+        # but to avoid reductions in atop, we always use args[0].ndim here, and fix it up below
+        out_ndim = arrays[0].ndim
+        atop_args = concat((a, tuple(range(out_ndim)[::-1])[:a.ndim]) for a in arrays)
+    else:
+        shapes = [getattr(arg, 'shape', ()) for arg in args]
+        shapes = [s if isinstance(s, Iterable) else () for s in shapes]
+        out_ndim = len(broadcast_shapes(*shapes))   # Raises ValueError if dimensions mismatch
```

Comment:
Just saw your comment, so I'm not sure if https://github.com/dask/dask/pull/2430/commits/3f50358a999753592c078ebb2ff983727a7cc406 addresses your concerns. That simply extracted the same sort of computations into their own functions.

Are you concerned about the special casing directly inside of `elemwise` because it's less clear, or because of performance? I'm not sure that https://github.com/dask/dask/pull/2430/commits/3f50358a999753592c078ebb2ff983727a7cc406 really helps with either of those...